code/controllers/subsystem/radiation.dm

```diff
@@ -10,83 +10,94 @@ SUBSYSTEM_DEF(radiation)
 
 /datum/controller/subsystem/radiation/fire(resumed)
 	while (processing.len)
-		var/datum/radiation_pulse_information/pulse_information = popleft(processing)
+		var/datum/radiation_pulse_information/pulse_information = processing[1]
 
 		var/datum/weakref/source_ref = pulse_information.source_ref
 		var/atom/source = source_ref.resolve()
 		if (isnull(source))
+			processing.Cut(1, 2)
 			continue
 
 		pulse(source, pulse_information)
 
 		if (MC_TICK_CHECK)
 			return
 
+		processing.Cut(1, 2)
+
 /datum/controller/subsystem/radiation/stat_entry(msg)
 	msg = "[msg] | Pulses: [processing.len]"
 	return ..()
 
 /datum/controller/subsystem/radiation/proc/pulse(atom/source, datum/radiation_pulse_information/pulse_information)
 	var/list/cached_rad_insulations = list()
+	var/list/cached_turfs_to_process = pulse_information.turfs_to_process
+	var/turfs_iterated = 0
+	for (var/turf/turf_to_irradiate as anything in cached_turfs_to_process)
+		turfs_iterated += 1
```

Comment:
It is worth noting that iterations in this form make a copy of the list, so they are not preferable if iterating and yielding over a LARGE list
This is not that tho, by large I mean like, 20k size. See the garbage subsystem.
That is why you'll see the while (thing.length) thing[thing.length] pattern so much though